nilearn/decomposition/_base.py

```diff
@@ -0,0 +1,762 @@
+        In new masker, pseudo number generator state used for random sampling.
+
+    Returns
+    ----------
+    returned_masker: _PCAMultiNiftiMasker or MultiNiftiMasker,
+        New masker with reduction capability
+    """
+    if not isinstance(masker, (NiftiMasker, MultiNiftiMasker)):
+        raise ValueError("Expected type: *NiftiMasker but got type %s" %
+                         type(masker))
+    params = masker.get_params()
+    if n_components is not None:
+        returned_masker = PCAMultiNiftiMasker(n_components=n_components,
+                                              random_state=random_state,
+                                              **params)
+    else:
+        returned_masker = MultiNiftiMasker(**params)
+    # If masker has already been fitted, intialize new masker with mask_img_
+    if hasattr(masker, 'mask_img_'):
+        returned_masker.mask_img = masker.mask_img_
+        returned_masker.fit()
+    return returned_masker
+
+
+def mask_and_reduce(masker, imgs, confounds=None,
+                    reduction_ratio='auto',
+                    n_components=None, random_state=None,
+                    memory_level=0,
+                    memory=Memory(cachedir=None),):
+        if isinstance(imgs, nibabel.Nifti1Image):
+            imgs = [imgs]
+
+        if reduction_ratio != 'auto':
+            reduction_ratio = float(reduction_ratio)
+            if reduction_ratio is None or reduction_ratio >= 1:
+                reduction_ratio = 1
+        # Precomputing number of samples for preallocation
+        subject_n_samples = np.zeros(len(imgs), dtype='int')
+
+        for i, img in enumerate(imgs):
+            if reduction_ratio == 'auto':
+                subject_n_samples[i] = min(n_components,
+                                           check_niimg_4d(img).get_shape()[3])
+            else:
+                subject_n_samples[i] = int(ceil(check_niimg_4d(img).
+                                           get_shape()[3] * reduction_ratio))
+        subject_limits = np.zeros(subject_n_samples.shape[0]+1,
+                                  dtype='int')
+        subject_limits[1:] = np.cumsum(subject_n_samples)
+        n_voxels = np.sum(masker.mask_img_.get_data())
+        n_samples = subject_limits[-1]
+
+        # We initialize data in memory or on disk
+        if n_voxels * n_samples * 8 > 2e9:
+            temp_dir = _get_dataset_dir('temp')
+            filename = os.path.join(temp_dir, 'data')
+            data = np.memmap(filename, dtype='float64', order='F', mode='w+',
+                             shape=(n_samples, n_voxels))
+        else:
+            temp_dir = None
+            data = np.empty((n_samples, n_voxels), order='F', dtype='float64')
+
+        if confounds is None:
+            confounds = [None] * len(imgs)
+
+        masker_type = type(masker)
+        for i, (img, confound) in enumerate(zip(imgs, confounds)):
+            this_data = cache(masker_type.transform, memory,
```

Comment:
transform method can cache the masking cleanly, is there any reason to do it by hand here?